base-notebook/start.sh

```diff
@@ -11,140 +12,185 @@ else
+    # - CHOWN_EXTRA: a comma separated list of paths to chown
+    # - CHOWN_HOME_OPTS / CHOWN_EXTRA_OPTS: arguments to the chown commands
 
-    # Only attempt to change the jovyan username if it exists
+    # Refit the jovyan user to the desired the user (NB_USER)
     if id jovyan &> /dev/null ; then
-        echo "Set username to: ${NB_USER}"
-        usermod -d "/home/${NB_USER}" -l "${NB_USER}" jovyan
+        if ! usermod --home "/home/${NB_USER}" --login "${NB_USER}" jovyan 2>&1 | grep "no changes" > /dev/null; then
+            echo "Updated the jovyan user:"
+            echo "- username: jovyan       -> ${NB_USER}"
+            echo "- home dir: /home/jovyan -> /home/${NB_USER}"
+        fi
+    elif ! id -u "${NB_USER}" &> /dev/null; then
+        echo "ERROR: Neither the jovyan user or '$NB_USER' exists."
+        echo "       This could be the result of stopping and starting, the"
+        echo "       container with a different NB_USER environment variable."
+        exit 1
+    fi
+    # Ensure the desired user (NB_USER) gets its desired user id (NB_UID) and is
+    # a member of the desired group (NB_GROUP, NB_GID)
+    if [ "${NB_UID}" != "$(id -u "${NB_USER}")" ] || [ "${NB_GID}" != "$(id -g "${NB_USER}")" ]; then
+        echo "Update ${NB_USER}'s UID:GID to ${NB_UID}:${NB_GID}"
+        # Ensure the desired group's existence
+        if [ "${NB_GID}" != "$(id -g "${NB_USER}")" ]; then
+            groupadd --force --gid "${NB_GID}" --non-unique "${NB_GROUP:-${NB_USER}}"
+        fi
+        # Recreate the desired user as we want it
+        userdel "${NB_USER}"
+        useradd --home "/home/${NB_USER}" --uid "${NB_UID}" --gid "${NB_GID}" --groups 100 --no-log-init "${NB_USER}"
     fi
 
-    # handle home and working directory if the username changed
+    # Move or symlink the jovyan home directory to the desired users home
+    # directory if it doesn't already exist, and update the current working
+    # directory to the new location if needed.
     if [[ "${NB_USER}" != "jovyan" ]]; then
-        # changing username, make sure homedir exists
-        # (it could be mounted, and we shouldn't create it if it already exists)
         if [[ ! -e "/home/${NB_USER}" ]]; then
-            echo "Copying home dir to /home/${NB_USER}"
+            echo "Attempting to copy /home/jovyan to /home/${NB_USER}..."
             mkdir "/home/${NB_USER}"
-            cp -a /home/jovyan/. "/home/${NB_USER}/" || ln -s /home/jovyan "/home/${NB_USER}"
+            if cp -a /home/jovyan/. "/home/${NB_USER}/"; then
+                echo "Success!"
+            else
+                echo "Failed!"
+                echo "Attempting to symlink /home/jovyan to /home/${NB_USER}..."
+                if ln -s /home/jovyan "/home/${NB_USER}"; then
+                    echo "Success!"
+                else
+                    echo "Failed!"
+                fi
+            fi
         fi
```

Comment:
@maresb Could you take https://github.com/jupyter/docker-stacks/issues/1478 into consideration here?

Thank you.

Reply:
@benz0li, thanks for pointing it out! I'm not quite ready to tackle this now, but I have added it to my to-do list above. I'm deliberately limiting this PR in scope to only a subset of #1052 in hopes of receiving a quick approval.

But with this PR in place, it should be easier to address #1478 since we'll already have the `if` statements.